message_ix/model/MESSAGE/model_core.gms

```diff
@@ -577,6 +590,59 @@ $macro COMMODITY_BALANCE(node,commodity,level,year,time) (
 * export from node and input into technologies located at 'location' taking from 'node' and 'time2' taking from 'time'
         - input(location,tec,vintage,year,mode,node,commodity,level,time2,time)                                        \
         * duration_time_rel(time,time2) * ACT(location,tec,vintage,year,mode,time2) )                                  \
+* +++++
+* commodity input and output associated with construction of new technology capacity (during vintage period)
+  + SUM( (location,tec)$( inv_tec(tec) AND map_tec(location,tec,year) ),                                               \
+* output by all new capacity of technologies located at 'location' sending to 'node' and 'time'
+        output_cap_new(location,tec,year,node,commodity,level,time)                                                    \
+        * CAP_NEW(location,tec,year)                                                                                   \
+* input by all new capacity of technologies located at 'location' taking from 'node' and 'time'
+        - input_cap_new(location,tec,year,node,commodity,level,time)                                                   \
+        * CAP_NEW(location,tec,year) )                                                                                 \
+* commodity input and output associated with retirement of technology capacity (via differentials of capacity of successive periods)
+* for first model period (differential with historical remaining capacity)
+  + SUM( (location,tec,vintage,year_all2)$( inv_tec(tec) AND map_tec_lifetime_extended(location,tec,vintage,year_all2)                  \
```

Comment:
The current formulation still does not treat commodity flows from historical vintages that have lifetimes that are not multiples of period length correctly. Those vintages only create a commodity flow in the first model period, but not in any later periods. In addition, the multiplication with `remaining_capacity_extended` implies that longer lifetime leads to higher flows although the opposite should happen, since longer lifetime means less retired capacity.

Reply:
The vintage years are still defined in the CAP variable if they are available in any model period. This makes sure that after the first model period all the capacity retirements are handeled correctly. The formualtion for the `remaining_capacity_extended` is revised.